src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -178,71 +176,83 @@ private void InitializeZeroTagPointIfNotInitialized()
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int length)
         {
-            int aggregatorIndex;
-            string[] seqKey = null;
+            var givenTags = new Tags(tagKeys, tagValues);
 
-            // GetOrAdd by TagKeys at 1st Level of 2-level dictionary structure.
-            // Get back a Dictionary of [ Values x Metrics[] ].
-            if (!this.keyValue2MetricAggs.TryGetValue(tagKeys, out var value2metrics))
+            if (!this.tagsToMetricPointIndexDictionary.TryGetValue(givenTags, out var aggregatorIndex))
             {
-                // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
-                seqKey = new string[length];
-                tagKeys.CopyTo(seqKey, 0);
-
-                value2metrics = new ConcurrentDictionary<object[], int>(ObjectArrayComparer);
-                if (!this.keyValue2MetricAggs.TryAdd(seqKey, value2metrics))
+                Tags sortedTags;
+                if (length > 1)
                 {
-                    this.keyValue2MetricAggs.TryGetValue(seqKey, out value2metrics);
-                }
-            }
+                    // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
+                    // Create a new array for the sorted Tag keys.
+                    var sortedTagKeys = new string[length];
+                    tagKeys.CopyTo(sortedTagKeys, 0);
 
-            // GetOrAdd by TagValues at 2st Level of 2-level dictionary structure.
-            // Get back Metrics[].
-            if (!value2metrics.TryGetValue(tagValues, out aggregatorIndex))
-            {
-                aggregatorIndex = this.metricPointIndex;
-                if (aggregatorIndex >= this.maxMetricPoints)
+                    // Create a new array for the sorted Tag values.
+                    var sortedTagValues = new object[length];
+                    tagValues.CopyTo(sortedTagValues, 0);
+
+                    Array.Sort(sortedTagKeys, sortedTagValues);
+
+                    sortedTags = new Tags(sortedTagKeys, sortedTagValues);
+                }
+                else
                 {
-                    // sorry! out of data points.
-                    // TODO: Once we support cleanup of
-                    // unused points (typically with delta)
-                    // we can re-claim them here.
-                    return -1;
+                    sortedTags = givenTags;
                 }
 
-                lock (value2metrics)
+                if (!this.tagsToMetricPointIndexDictionary.TryGetValue(sortedTags, out aggregatorIndex))
```

Comment:
minor optimization - this is now executed for keys.lengtgh==1 also. we can avoid dual look up with same tags!

Reply:
I have updated to code to not do an additional lookup if tags `tagsLength <= 1`.

Reply:
Could you update the PR description with the final code, along with updated benchmarks/stress test before/after?

Reply:
The Benchmarks/Stress Test numbers are the same as:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2805#discussion_r794096208
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2805#discussion_r794101246

The numbers didn't change because of the optimizations of the last commit. That's because those optimizations would only show their effect when the tags are not yet added to the dictionary. Once, the tags are added they will not touch the paths that we have optimized. The `GetHashCode` optimization didn't affect the numbers much either.

Reply:
Consider adding a changelog entry - Performance improvements. (Users are advised to provide tags with same Key order, to achieve maximum performance)..

Reply:
I have updated the Benchmark and Stress Test numbers.